src/node.cpp

```diff
@@ -941,6 +972,20 @@ bool Node::setparent(Node* p)
         return false;
     }
 
+    node_counter nc;
+    bool gotnc = false;
+
+    Node *originalancestor = firstancestor();
+    handle oah = originalancestor->nodehandle;
+    if (oah == client->rootnodes[0] || oah == client->rootnodes[1] || oah == client->rootnodes[2] || originalancestor->inshare)
+    {
+        nc = subnodeCounts();
+        gotnc = true;
+
+        // nodes moving from cloud drive to rubbish for example.  Nodes should not move between inshares and owned nodes
```

Comment:
Should they be copy&delete instead of move? (I don't know by heart)

Reply:
I checked with Jon, and in fact they can move between inshares, when the inshares are from the same user.  However the changes are sent as a delete of that node handle, followed by a node tree that contains that same node handle.   The delete processing just sets a `changed.deleted` flag on the node for that handle, that is unset when processing the node tree, followed by setparent for each received node.    So the code will behave correctly, and I'll fix the comment.  thanks